clusterctl/clusterdeployer/clusterdeployer.go

```diff
@@ -1,38 +1,392 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
 package clusterdeployer
 
 import (
 	"fmt"
 
-	"sigs.k8s.io/cluster-api/errors"
+	"github.com/golang/glog"
+	"io/ioutil"
+	"os"
+	clustercommon "sigs.k8s.io/cluster-api/pkg/apis/cluster/common"
 	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	"sigs.k8s.io/cluster-api/util"
+	"time"
 )
 
+// Provider specific logic. Logic here should eventually be optional & additive.
+type ProviderDeployer interface {
+	// TODO: This requirement can be removed once after: https://github.com/kubernetes-sigs/cluster-api/issues/158
+	GetIP(cluster *clusterv1.Cluster, machine *clusterv1.Machine) (string, error)
+	// TODO: This requirement can be removed after: https://github.com/kubernetes-sigs/cluster-api/issues/160
+	GetKubeConfig(cluster *clusterv1.Cluster, master *clusterv1.Machine) (string, error)
+}
+
 // Can provision a kubernetes cluster
 type ClusterProvisioner interface {
 	Create() error
 	Delete() error
 	GetKubeconfig() (string, error)
 }
 
+// Provides interaction with a cluster
+type ClusterClient interface {
+	Apply(string) error
+	WaitForClusterV1alpha1Ready() error
+	GetClusterObjects() ([]*clusterv1.Cluster, error)
+	GetMachineObjects() ([]*clusterv1.Machine, error)
+	CreateClusterObject(*clusterv1.Cluster) error
+	CreateMachineObjects([]*clusterv1.Machine) error
+	UpdateClusterObjectEndpoint(string) error
+	Close() error
```

Comment:
Eventually, will there be Delete methods for Machine and Cluster? Is that out of scope for clusterctl?

Reply:
Deletion is in scope for clusterctl. 

If that ability is added to the ClusterDeployer, it would make sense to extend the ClusterClient interface to handle the needed methods at that point.